README
Lemonldap-NG
====================

Lemonldap::NG is a modular Web-SSO based on Apache::Session modules. It
simplifies the build of a protected area with a few changes in the application.

It manages both authentication and authorization and provides headers for
accounting. So you can have a full AAA protection for your web space as
described below.

 1 - Installation
 2 - Authentication, Authorization and Accounting mechanisms
     2.1 - Authentication
     2.2 - Authorization
     2.3 - Accounting
 3 - Session storage system
 4 - Logout system
 5 - Author
 6 - Copyright and licence

1 - INSTALLATION
================

Lemonldap::NG is a different project than Lemonldap and contains all you need
to use and administer it. So softwares, like Lemonldap webmin module, may not
work with Lemonldap::NG.

The Apache module part (Lemonldap::NG::Handler) works both with Apache 1.3.x
and 2.x ie mod_perl 1 and 2 (but not with mod_perl 1.99). Portal and Manager
act as CGI, so they can work everywhere.

See INSTALL file in the whole source tree of Lemonldap::NG for a complete
installation documentation.

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

2 - AUTHENTICATION, AUTHORIZATION AND ACCOUNTING MECHANISMS
===========================================================

Warning: Lemonldap::NG configuration has to be edited using the manager unless
you know exactly what you are doing. The parameters discussed here are all in
the configuration tree.

2.1 - Authentication

If a user isn't authenticated and attemps to connect to an area protected by a
Lemonldap::NG compatible handler, he is redirected to a portal. The portal
authenticates user with a ldap bind by default, but you can also use another
authentication sheme like using x509 user certificates (see
Lemonldap::NG::Portal::AuthSSL(3) for more).

Lemonldap::NG use session cookies generated by Apache::Session so as secure as a
128-bit random cookie. You may use the securedCookie options to avoid session
hijacking.

You have to manage life of sessions by yourself since Lemonldap::NG knows
nothing about the L<Apache::Session> module you've choosed, but it's very easy
using a simple cron script because Lemonldap::NG::Portal stores the start
time in the _utime field.
By default, a session stay 10 minutes in the local storage, so in the worth
case, a user is authorized 10 minutes after he lost his rights.

2.2 - Authorization

Authorization is controled only by handlers because the portal knows nothing
about the way the user will choose. When configuring your Web-SSO, you have to:

 * choose the ldap attributes you want to use to manage accounting and
   authorization.
 * create Perl expressions to define user groups (using ldap attributes)
 * create an array foreach virtual host associating URI regular expressions and
   Perl expressions to use to grant access.

Example (See Lemonldap::NG::Common::Conf(3) to see how configuration is stored)

 * Exported variables :

   # Custom-Name    => LDAP attribute
     cn             => cn
     departmentUID  => departmentUID
     login          => uid

 * User groups :

   # Custom-Name => group definition
     group1      => { $departmentUID eq "unit1" or $login = "xavier.guimard" }

 * Area protection:

  # Each VirtualHost has its own configuration
  # associating URL regexp to Perl expression
    * www1.domain.com :
          ^/protected/.*$ => $groups =~ /\bgroup1\b/
          default         => accept
      },
    * www2.domain.com => {
          ^/site/.*$       => $uid eq "admin" or $groups =~ /\bgroup2\b/
          ^/(js|css)       => accept
          default          => deny
      },
  },

2.2.1 - Performance

You can use Perl expressions as complicated as you want and you can use all
the exported LDAP attributes (and create your own attributes: with 'macros'
mechanism) in groups evaluations, area protections or custom HTTP headers
(you just have to call them with a "$").

You have to be careful when choosing your expressions:

 * groups and macros are evaluated each time a user is redirected to the portal
 * virtual host rules and exported headers are evaluated for each request on a
   protected area.

It is also recommanded to use the groups mechanism to avoid having to evaluate
a long expression at each HTTP request:

  # Virtual hosts :
      ...
      www1.domain.com :
          ^/protected/.*$   => $groups =~ /\bgroup1\b/

You can also use LDAP filters, or Perl expression or mixed expressions in
groups definitions. Perl expressions has to be enclosed with {}:

 * group1 => (|(uid=xavier.guimard)(ou=unit1))
 * group1 => {$uid eq "xavier.guimard" or $ou eq "unit1"}
 * group1 => (|(uid=xavier.guimard){$ou eq "unit1"})

It is also recommanded to use Perl expressions to avoid requiering the LDAP
server more than 2 times per authentication.

2.3 - Accounting

2.3.1 - Logging portal access>

Lemonldap::NG::Portal doesn't log anything by default, but it's easy to
overload log method for normal portal access.

2.3.2 - Logging application access

Because a Web-SSO knows nothing about the protected application, it can't do
more than logging URL. As Apache does this fine, L<Lemonldap::NG::Handler>
gives it the name to used in logs. The whatToTrace parameter indicates
which variable Apache has to use ($uid by default).

The real accounting has to be done by the application itself which knows the
result of SQL transaction for example.

Lemonldap::NG can export HTTP headers either using a proxy or protecting
directly the application. By default, the Auth-User field is used but you can
change it using the exportedHeaders parameters (in the Manager, each virtual
host as custom headers branch). This parameters contains an associative array
per virtual host:

 * keys are the names of the choosen headers
 * values are Perl expressions where you can use user datas stored in the
   global storage.

Example:

 * www1.domain.com :
        Auth-User     => $uid
        Unit          => $ou
 * www2.domain.com :
        Authorization => "Basic ".encode_base64($employeeNumber.":dummy")
        Remote-IP     => $ip

3 - SESSION STORAGE SYSTEM

Lemonldap::NG use 3 levels of cache for authenticated users:

 * an Apache::Session::* module used by lemonldap::NG::Portal to store
   authenticated user parameters,
 * a Cache::Cache* module used by Lemonldap::NG::Handler to share authenticated
   users between Apache's threads or processus and of course between virtual
   hosts on the same machine
 * Lemonldap::NG::Handler variables : if the same user use the same thread or
   processus a second time, no request are needed to grant or refuse access.
   This is very efficient with HTTP/1.1 Keep-Alive system.

So the number of request to the central storage is limited to 1 per active
user each 10 minutes.

Lemonldap::NG is very fast, but you can increase performance using a 
Cache::Cache module that does not use disk access.

4 - LOGOUT SYSTEM

Lemonldap::NG provides a single logout system: you can use it by adding a link
to the portal with "logout=1" parameter in the portal (See
Lemonldap::NG::Portal) and/or by configuring handler to intercept some URL
(See Lemonldap::NG::Handler) directly in the manager interface or in apache
configuration file. The logout system:
 * delete session in the global session storage,
 * replace Lemonldap::NG cookie by '',
 * delete handler caches only if logout action was started from a protected
   application and only in the current Apache server. So in other servers,
   session is still in cache for 10 minutes maximum if the user was connected
   on it in the last 10 minutes.

5 - AUTHOR

Xavier Guimard, x.guimard@free.fr

6 - COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 by Xavier Guimard x.guimard@free.fr

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Lemonldap::NG::Handler.

0.91  Mon Jun 29 11:54:02 2009
	- CDA in core
	- AuthBasic now use SOAP
	- POST requests generation (used to post login/password in non
	  compatible applications)
	- logs written in PerlLogHandler
	- Sympa auto login handler
	- Double session mechanism (1 secured and the other not)

0.9   Thu Dec 25 8:43:55 2008
	- Configuration now managed by Lemonldap::NG::Common::Conf
	- Custom functions
	- Status now couns only past minutes, not current one
	- Protection management in Handler/CGI

0.89  Mon Aug 25 21:59:34 2008
	- Bug in default condition: rules are always "deny" or "accept"
	- More examples
	- New file : AuthBasic.pm

0.88  Fri jun 6 14:58:41 2008
	- Status.pm test failed in jail

0.87  Tue may 9 22:32:44 2008
	- Remove Apache2::compat dependency
	- New status module

0.86  Mon apr 7 14:52:30 2008
	- logout bug : logout_sso target was not running (Closes: #308856 /
	  forge.objectweb.org)

0.85  Fri feb 8 17:50:25 2008
	- local overload was not taken in account
	- little bugs in redirections

0.84  Mon Jul 30 21:26:19 2007
	- New parameter to fix port in redirections.

0.83  Sun Jul 22 22:19:12 2007
	- New logout system : URL can now be declared in Manager
	  interface

0.82  Fri Jun 1 6:52:43 2007
	- Little bug: SharedConf.pm needs Manager::Conf and not Manager

0.81  Fri Apr 15 7:15:56 2007
	- More tests

0.8   Thu Apr 12 7:08:08 2007
	- Little bug in unprotect function (Handler::Simple)
	- New feature: Lemonldap::NG::Handler::CGI

0.77  Wed Mar 22 7:18:48 2007
	- logout system
	- $date is now available in rules
	- Closes bug #306761 / forge.objectweb.org

0.76  Fri Mar 9 7:14:42 2007
	- Little bug correction
	- Delete of DBI.pm which is not more in use

0.75  Sat Feb 24 16:36:56 2007
	- Adding cross-domain-authentication support
	- local configuration overrides global configuration

0.74  Sun Feb 4 19:27:34 2007
	- unprotect system documentation
	- remove warning on Apache-1.3 ($ENV{MOD_PERL_API_VERSION} does not exist)

0.73  Thu Jan 4 15:34:22 2007
	- In VirtualHost context, PerlInitHandler is an alias for
	  PerlPostReadRequestHandler. So by encoding characters (%40 for 'a'),
	  users can bypass some restrictions.
	  Now PerlInitHandler is replaced by PerlHeaderParserHandler, so
	  Lemonldap::NG is called after URI translation phase.
	- Correction of a bug in internal redirections: now internal
	  redirections are not examined: for example, http://test.example.com/
	  is internaly redirected to /index.pl, but only the first request (/)
	  is tested.

0.72  Thu Jan 4 9:24:18 2007
	- Correction in apache-1.3 configuration example file

0.71  Sun Dec 31 13:47:49 2006
	- Bug corrections in Vhosts: default header does not work
	                             Safe does not work because lmSetHeaderIn
	                                   was not shared
	- Example correction

0.7   Tue Dec 19 19:09:09 2006
	- config is now shared with Lemonldap::NG::Manager::Conf
	- use Safe for locationRules and exportedHeaders 'eval'

0.63  Sun Dec 17 14:54:14 2006
	- bug corrections

0.621 Tue Dec 5 20:12:12 2006
	- move Handler.pod to Handler.pm to avoid CPAN bug
	- remove mod_perl dependency

0.62  Fri Nov 3 8:08:37 2006
	- correction in documentation
	- add mod_perl in PREREQ_PM (CPAN tester failed)

0.61  Thu Nov 2 16:53:03 2006
	- little correction in documentaion
	- wrong version number in Handler/SharedConf/DBI.pm

0.6   Thu Nov 2 15:16:56 2006
	- Bug correction in Proxy.pm
	- New documentation
	- Handler.pm becomes Handler/Simple.pm
	- Tested with 40.000 users

0.5   Tue Oct 17 15:35:28 2006
	- Bug correction in Proxy.pm
	- First running release under Apache2

0.4   Mon Oct 17 15:00:20 2006
	- Bug corrections in Apache2 support

0.3   Sat Oct 14 14:54:56 2006
	- Bug correction in both Apache 1 and 2 support

0.2   Mon Oct 9 17:32:20 2006
	- Introducing Apache2 support

0.11  Fri Oct 6 17:24:00 2006
	- bugs corrections

0.1   Thu Oct 5 19:00:00 2006
	- new package: Lemonldap::NG::Handler::SharedConf::DBI including
	a complete shared system based on a shared DB

0.06  Thu Jul 13 19:21:00 2006
	- comments and little correction in Handler.pm
	- bug correction in SharedConf.pm

0.05  Fri Jul 29 17:32:00 2005
	- bug correction in Vhost.pm

0.04  Fri Jul 29 16:02:00 2005
	- bug correction in SharedConf.pm

0.03  Wed Jul 27 21:00:00 2005
	- bug correction in Handler.pm: &conditionSub returned nothing except
	for default configuration
	- modification in SharedConf.pm : adding (undocumented)
	functionnalities to be able to dinamically update configuration

0.02  Sat Jul 2 10:35:00 2005
	- bug corrections
	- Vhost.pm modification: add header personalization per virtual host

0.01  Thu Jun 23 17:36:59 2005
	- original version; created by h2xs 1.23 with options
		-AXn Lemonldap::NG::Handler



lib/Lemonldap/NG/Handler.pm
## @file
# Alias for Lemonldap::NG::Handler::SharedConf

## @class
# Alias for Lemonldap::NG::Handler::SharedConf
package Lemonldap::NG::Handler;

our $VERSION = "0.91";
use Lemonldap::NG::Handler::SharedConf;
use base qw(Lemonldap::NG::Handler::SharedConf);

1;

__END__

=pod

=head1 NAME

Lemonldap::NG::Handler - The Apache protection module part of
Lemonldap::NG Web-SSO system.

=head1 SYNOPSIS

=head2 Create your Apache module

Create your own package (example using a central configuration database):

  package My::Package;
  use Lemonldap::NG::Handler::SharedConf;
  @ISA = qw(Lemonldap::NG::Handler::SharedConf);
  
  __PACKAGE__->init ( {
    # Local storage used for sessions and configuration
    localStorage        => "Cache::DBFile",
    localStorageOptions => {...},
    # How to get my configuration
    configStorage       => {
        type                => "DBI",
        dbiChain            => "DBI:mysql:database=lemondb;host=$hostname",
        dbiUser             => "lemonldap",
        dbiPassword         => "password",
    }
    # Uncomment this to activate status module
    # status                => 1,
  } );

=head2 Configure Apache

Call your package in /apache-dir/conf/httpd.conf:

  # Load your package
  PerlRequire /My/File
  # TOTAL PROTECTION
  PerlHeaderParserHandler My::Package
  # OR SELECTED AREA
  <Location /protected-area>
    PerlHeaderParserHandler My::Package
  </Location>

The configuration is loaded only at Apache start. Create an URI to force
configuration reload, so you don't need to restart Apache at each change:

  # /apache-dir/conf/httpd.conf
  <Location /location/that/I/ve/choosed>
    Order deny,allow
    Deny from all
    Allow from my.manager.com
    PerlHeaderParserHandler My::Package->refresh
  </Location>
  
You can also unprotect an URI

  <Files "*.gif">
    PerlHeaderParserHandler My::Package->unprotect
  </Files>

To display the status page, add something like this :

  <Location /status>
    Order deny,allow
    Allow from 10.1.1.0/24
    Deny from all
    PerlHeaderParserHandler My::Package->status
  </Location>

If your application has a "logout" URL, you can configure it directly in Apache
configuration file (or in the manager interface). THIS IS DEPRECATED, use the
manager :

  <Location /logout>
    PerlHeaderParserHandler My::Package->logout
  </Location>

=head1 DESCRIPTION

Lemonldap::NG is a modular Web-SSO based on Apache::Session modules. It
simplifies the build of a protected area with a few changes in the application.

It manages both authentication and authorization and provides headers for
accounting. So you can have a full AAA protection for your web space as
described below.

The Apache module part works both with Apache 1.3.x and 2.x ie mod_perl 1 and 2
but B<not with mod_perl 1.99>.

=head2 Authentication, Authorization, Accounting

=head3 B<Authentication>

If a user isn't authenticated and attemps to connect to an area protected by a
Lemonldap::NG compatible handler, he is redirected to a portal. The portal
authenticates user with a ldap bind by default, but you can also use another
authentication sheme like using x509 user certificates (see
L<Lemonldap::NG::Portal::AuthSSL> for more).

Lemonldap::NG use session cookies generated by L<Apache::Session> so as secure
as a 128-bit random cookie. You may use the C<securedCookie> options of
L<Lemonldap::NG::Portal> to avoid session hijacking.

You have to manage life of sessions by yourself since Lemonldap::NG knows
nothing about the L<Apache::Session> module you've choosed, but it's very easy
using a simple cron script because L<Lemonldap::NG::Portal> stores the start
time in the C<_utime> field.
By default, a session stay 10 minutes in the local storage, so in the worth
case, a user is authorized 10 minutes after he lost his rights.

=head3 B<Authorization>

Authorization is controled only by handlers because the portal knows nothing
about the way the user will choose. When configuring your Web-SSO, you have to:

=over

=item * choose the ldap attributes you want to use to manage accounting and
authorization (see C<exportedHeaders> parameter in L<Lemonldap::NG::Portal>
documentation).

=item * create Perl expressions to define user groups (using ldap attributes)

=item * create an array foreach virtual host associating URI regular
expressions and Perl expressions to use to grant access.

=back

=head4 Example (See L<Lemonldap::NG::Manager> to see how configuration is
stored)

Exported variables (values will be stored in session database by
L<Lemonldap::NG::Portal>):

  exportedVars => {
      cn            => "cn",
      departmentUID => "departmentUID",
      login         => "uid",
  },

User groups (values will be stored in session database by
L<Lemonldap::NG::Portal>):

  groups => {
      group1 => '{ $departmentUID eq "unit1" or $login = "xavier.guimard" }',
      ...
  },

Area protection:

  locationRules => {
      www1.domain.com => {
          '^/protected/.*$' => '$groups =~ /\bgroup1\b/',
          default           => 'accept',
      },
      www2.domain.com => {
          '^/site/.*$' => '$uid eq "xavier.guimard" or $groups =~ /\bgroup2\b/',
          '^/(js|css)' => 'accept',
          default      => 'deny',
      },
  },

=head4 Performance

You can use Perl expressions as complicated as you want and you can use all
the exported LDAP attributes (and create your own attributes: with 'macros'
mechanism. See L<Lemonldap::NG::Manager>) in groups evaluations, area
protections or custom HTTP headers (you just have to call them with a "$").

You have to be careful when choosing your expressions:

=over

=item * C<groups> and C<macros> are evaluated each time a user is redirected to
the portal,

=item * C<locationRules> and C<exportedheaders> are evaluated for each request
on a protected area.

=back

It is also recommanded to use the C<groups> mechanism to avoid having to
evaluate a long expression at each HTTP request:

  locationRules => {
      www1.domain.com => {
          '^/protected/.*$' => '$groups =~ /\bgroup1\b/',
      },
  },

You can also use LDAP filters, or Perl expression or mixed expressions in
C<groups> parameter. Perl expressions has to be enclosed with C<{}>:

=over

=item * C<group1 =E<gt> '(|(uid=xavier.guimard)(ou=unit1))'>

=item * C<group1 =E<gt> '{$uid eq "xavier.guimard" or $ou eq "unit1"}'>

=item * C<group1 =E<gt> '(|(uid=xavier.guimard){$ou eq "unit1"})'>

=back

It is also recommanded to use Perl expressions to avoid requiering the LDAP
server more than 2 times per authentication.

=head3 B<Accounting>

=head4 I<Logging portal access>

L<Lemonldap::NG::Portal> doesn't log anything by default, but it's easy to overload
C<log> method for normal portal access or using C<error> method to know what
was wrong if C<process> method has failed.

=head4 I<Logging application access>

Because an handler knows nothing about the protected application, it can't do
more than logging URL. As Apache does this fine, L<Lemonldap::NG::Handler>
gives it the name to used in logs. The C<whatToTrace> parameters indicates
which variable Apache has to use (C<$uid> by default).

The real accounting has to be done by the application itself which knows the
result of SQL transaction for example.

Lemonldap::NG can export HTTP headers either using a proxy or protecting
directly the application. By default, the C<Auth-User> field is used but you
can change it using the C<exportedHeaders> parameters (stored in the
configuration database). This parameters contains an associative array per
virtual host:

=over

=item * B<keys> are the names of the choosen headers

=item * B<values> are Perl expressions where you can use user datas stored in
the global store by calling them C<$E<lt>varnameE<gt>>.

=back

Example:

  exportedHeaders => {
      www1.domain.com => {
          'Auth-User' => '$uid',
          'Unit'      => '$ou',
      },
      www2.domain.com => {
          'Authorization' => '"Basic ".encode_base64($employeeNumber.":dummy")',
          'Remote-IP'     => '$ip',
      },
  }

=head2 Session storage systems

Lemonldap::NG use 3 levels of cache for authenticated users:

=over

=item * an Apache::Session::* module choosed with the C<globalStorage>
parameter (completed with C<globalStorageOptions>) and used by
L<lemonldap::NG::Portal> to store authenticated user parameters,

=item * a L<Cache::Cache> module choosed with the C<localStorage> parameter
(completed with C<localStorageOptions>) and used to share authenticated users
between Apache's threads or processus and of course between virtual hosts,

=item * Lemonldap::NG::Handler variables: if the same user use the same thread
or processus a second time, no request are needed to grant or refuse access.
This is very efficient with HTTP/1.1 Keep-Alive system.

=back

So the number of request to the central storage is limited to 1 per active
user each 10 minutes.

Lemonldap::NG is very fast, but you can increase performance using a
L<Cache::Cache> module that does not use disk access.

=head2 Logout system

Lemonldap::NG provides a single logout system: you can use it by adding a link
to the portal with "logout=1" parameter in the portal (See
L<Lemonldap::NG::Portal>) and/or by configuring handler to intercept some URL
(See Sinopsys). The logout system:

=over

=item * delete session in the global session storage,

=item * replace Lemonldap::NG cookie by '',

=item * delete handler caches only if logout action was started from a
protected application and only in the current Apache server. So in other
servers, session is still in cache for 10 minutes maximum if the user was
connected on it in the last 10 minutes.

=back

You can also configure rules in the Manager interface to intercept logout URL.
See L<Lemonldap::NG::Manager> and L<Lemonldap::NG::Handler> for more.

=head1 USING LEMONLDAP::NG::HANDLER FOR DEVELOPMENT

Lemonldap::NG::Handler provides different modules:

=over

=item * L<Lemonldap::NG::Handler::Simple>: base module. It can be used
directly to protect a single host.

=item * L<Lemonldap::NG::Handler::Vhost>: module used to managed virtual hosts.

=item * L<Lemonldap::NG::Handler::SharedConf>: with this module, the
configuration can be centralized. Inherits from
L<Lemonldap::NG::Handler::Vhost> and L<Lemonldap::NG::Handler::Simple>.

=item * L<Lemonldap::NG::Handler::CGI>: if you have only a few Perl CGI to
protect, you can use this module in your CGI instead of protecting it under
L<Lemonldap::NG::Handler::SharedConf>.

=item * L<Lemonldap::NG::Handler::Proxy>: this module isn't used to manage
security but is written to create a reverse-proxy without using mod_proxy. In
some case, mod_proxy does not manage correctly some redirections, that is why
this module still exists.

=back

All those modules are compatible both with Apache and mod_perl version 1 and 2,
but NOT with mod_perl 1.99. If you use Linux distributions like Debian Sarge
who provide mod_perl 1.99 for Apache2, you have to use Apache-1.3 or to
download a mod_perl2 backport.

=head1 SEE ALSO

L<Lemonldap::NG::Handler::SharedConf>,
L<Lemonldap::NG::Portal>, L<Lemonldap::NG::Manager>,
L<http://wiki.lemonldap.objectweb.org/xwiki/bin/view/NG/Presentation>

=head1 AUTHOR

Xavier Guimard, E<lt>x.guimard@free.frE<gt>

=head1 BUG REPORT

Use OW2 system to report bug or ask for features:
L<http://forge.objectweb.org/tracker/?group_id=274>

=head1 DOWNLOAD

Lemonldap::NG is available at
L<http://forge.objectweb.org/project/showfiles.php?group_id=274>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 by Xavier Guimard E<lt>x.guimard@free.frE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.

Lemonldap was originaly written by Eric German who decided to publish him in
2003 under the terms of the GNU General Public License version 2.
Lemonldap::NG is a complete rewrite of Lemonldap and is able to have different
policies in a same Apache virtual host.

=cut
